### PR TITLE
Provide single Object base class

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR provides a single Object base class (largely the New Mapping one) and refactors ObjectAdapter and ServantLocator.

I don't plan to using/improving SharedPtr further after this PR.

### CODE REVIEW

File: cpp/include/Ice/ObjectAdapter.h [hunk 1]
```diff
@@ -343,7 +379,7 @@ class ICE_CLASS(ICE_API) ObjectAdapter
      * @see #find
      * @see #findFacet
      */
-    virtual ::std::shared_ptr<::Ice::Object> findByProxy(const ::std::shared_ptr<ObjectPrx>& proxy) const = 0;
+    virtual ::std::shared_ptr<::Ice::Object> findByProxy(const ObjectPrxPtr& proxy) const = 0;
```
Comment (externl) [inline cpp/include/Ice/ObjectAdapter.h:382]:
```suggestion
    virtual ::std::shared_ptr<Object> findByProxy(const ObjectPrxPtr& proxy) const = 0;
```

File: cpp/include/Ice/OutgoingAsync.h [hunk 2]
```diff
@@ -682,7 +682,7 @@ class AsyncCallback : public GenericCallbackBase
     AsyncCallback(const TPtr& instance, Callback cb, Callback sentcb = 0) :
         _callback(instance), _completed(cb), _sent(sentcb)
     {
-        checkCallback(instance, cb != 0);
+        checkCallback(instance != 0, cb != 0);
```
Comment (externl) [inline cpp/include/Ice/OutgoingAsync.h:685]: We should try to use `nullptr`.
```suggestion
        checkCallback(instance != nullptr, cb != nullptr);
```

Comment (externl) [inline cpp/include/Ice/OutgoingAsync.h:685]: We should try you use `nullptr`
```suggestion
        checkCallback(instance != nullptr, cb != nullptr);
```

File: cpp/include/Ice/Instrumentation.h [hunk 3]
```diff
@@ -385,7 +385,7 @@ class ICE_CLASS(ICE_API) CommunicatorObserver
      * interface can use this object to get the Ice run-time to re-obtain observers for observed objects.
      * @param updater The observer updater object.
      */
-    virtual void setObserverUpdater(const ObserverUpdaterPtr& updater) = 0;
+    virtual void setObserverUpdater(const  ObserverUpdaterPtr& updater) = 0;
```
Comment (pepone) [inline cpp/include/Ice/Instrumentation.h:388]:
```suggestion
    virtual void setObserverUpdater(const ObserverUpdaterPtr& updater) = 0;
```

**Review by externl — APPROVED** [review]
Looks good to me!

**Review by pepone — APPROVED** [review]

File: cpp/include/Ice/OutgoingAsync.h (hunk 2 above)
Comment (bernardnormier) [reply cpp/include/Ice/OutgoingAsync.h:685]: I tried, it doesn't work.

**Pull request merged**